community/tutorials/gdnative/gdnative-c-example.rst

```diff
@@ -0,0 +1,338 @@
+
+The below code includes our header files that we need and then defines two pointers to two different structs. 
+GDNative supports a large collection for functions for calling back into the main Godot executable. In order for your module to have access to these functions, GDNative provides your application with a struct containing pointers to all these functions.
+
+To keep this implementation modular and easily extendable, the core functions are available directly through the "core" API struct, but additional functions have their own "GDNative structs" that are accessible through extensions. 
+
+In our example, we access one of these extension to gain access to the functions specifically needed for NativeScript.
+
+.. code:: C
+
+    #include <gdnative_api_struct.gen.h>
+    #include <stdio.h>
+    #include <stdlib.h>
+    #include <string.h>
+
+    const godot_gdnative_core_api_struct *api = NULL;
+    const godot_gdnative_ext_nativescript_api_struct *nativescript_api = NULL;
+
+A NativeScript behaves like any other script in Godot. Because the NativeScript API is rather low level, it requires the library to specify many things more verbosely than other scripting systems, such as GDScript. When a NativeScript instance gets created, a library-given constructor gets called. When that instance gets destroyed, the given destructor will be executed.
+
+These are forward declarations for the functions we'll be implementing for our object. A constructor and destructor is needed. Additionally, the object will have a single method called get_data.
+
+.. code:: C
+
+    void *simple_constructor(godot_object *p_instance, void *p_method_data);
+    void simple_destructor(godot_object *p_instance, void *p_method_data, void *p_user_data);
+    godot_variant simple_get_data(godot_object *p_instance, void *p_method_data
+        , void *p_user_data, int p_num_args, godot_variant **p_args);
+
+Next up is the first of the entry points Godot will call when our dynamic library is loaded. These methods are all prefixed with godot (you can change this later on) followed by their name. ``gdnative_init`` is a function that initialises our dynamic library. Godot will give it a pointer to a structure that contains various bits of information we may find useful amongst which the pointers to our API structures.
+
+For any additional API structures we need to loop through our extensions array and check the type of extension.
+
+.. code:: C
+
+    void GDN_EXPORT godot_gdnative_init(godot_gdnative_init_options *p_options) {
+        api = p_options->api_struct;
+
+        // now find our extensions
+        for (int i = 0; i < api->num_extensions; i++) {
+            switch (api->extensions[i]->type) {
+                case GDNATIVE_EXT_NATIVESCRIPT: {
+                    nativescript_api = (godot_gdnative_ext_nativescript_api_struct *)api->extensions[i];
+                }; break;
+                default: break;
+            }
+        }
+    }
+
```

Comment:
I'd simplify that to
```
void GDN_EXPORT godot_gdnative_init(godot_gdnative_init_options *p_options) {

    api = p_options->api_struct;

    // now find our extension
    for (int i = 0; i < api->num_extensions; i++) {
        if (api->extensions[i]->type == GDNATIVE_EXT_NATIVESCRIPT) {
            nativescript_api = (godot_gdnative_ext_nativescript_api_struct *)
                api->extensions[i];
            break;
        }
    }  
}
```

Reply:
No, I've left that case on purpose, if  you use this as a boiler plate to write other GDNative modules it is very likely you'll use any of the other extensions as well. Case makes more sense in that.

I think even for the C sample in GDNative demos it may be worth just getting all extensions and adding version checking to them with a remark that you don't have to include the APIs you won't want but at least you can see they exist.

Reply:
I'm more freaked out by the `};` 😛

Reply:
I hate boiler plates and would only add that to where it belongs, in the next expanded example, but that's just my personal opinion, you can forget it :)

Reply:
@karroffel geez, I'm I really so old fashioned that I actually feel without the ; it's missing something?

@capnm I wouldn't add it to the tutorial, but the switch does hint at other extensions being there. I might however be tempted adding it to the gdnative-demos repo :)